Suppress the "read-modify-write" error if type is a struct or a class

ErnyTech opened the pull request:
If a struct or class with shared type implements the opUnary or opOpAssign
operator then it is possible to perform "read-modify-write" operations
because the operator's implementation is supposed to perform atomic operations.

The purpose of this modification is to allow the creation of wrappers
(with structs or classes) to run atomic operations silently, for example:

```
shared struct Atomic {
    int a;

    int opUnary(string s)() if (s == "++")
    {
        import core.atomic : atomicOp;
        return atomicOp!"+="(a, 1);
    }
}

Atomic atomicvar;
atomicvar++; // Safe! Atomic struct implements opUnary
```

Geod24 commented on src/dmd/expression.d:
```diff
@@ -1353,7 +1353,7 @@ extern (C++) abstract class Expression : ASTNode
     {
         //printf("Expression::checkReadModifyWrite() %s %s", toChars(), ex ? ex.toChars() : "");
-        if (!type || !type.isShared())
+        if (!type || !type.isShared() || type.isTypeStruct() || type.isTypeClass())
```
That doesn't look right. It's blindly allowing `struct` / `class`

ErnyTech replied:
If the class/struct does not implement the operator the compiler will give the error "is not a scalar" (https://github.com/dlang/dmd/blob/d9826c828a8b228713f71136dc6108c42e295352/src/dmd/expressionsem.d#L5574)

So I don't think further checks are necessary

Geod24 reviewed (dismissed):
No test

ErnyTech commented:
Added compilable test

thewilsonator commented:
```
Searching for tabs
./changelog/pending.dd:321:	int opUnary(string s)() if (s == "++")
./changelog/pending.dd:322:	{
./changelog/pending.dd:323:		import core.atomic : atomicOp;
./changelog/pending.dd:324:		return atomicOp!"+="(a, 1);
./changelog/pending.dd:325:	}
```

Geod24 commented on test/compilable/readmodify_structclass.d:
```diff
@@ -0,0 +1,35 @@
+{
+    S s;
+    s++;
+    shared(C) c = new C();
```
Why does this need to be shared if `C` is a `shared class` ?

ErnyTech replied:
Using shared in the class declaration does not seem to make the class type "implicitly" shared unlike what it does in a struct

Commit:
Suppress the "read-modify-write" error if type is a struct or a class

If a struct or class with shared type implements the opUnary or opOpAssign
operator then it is possible to perform "read-modify-write" operations
because the operator's implementation is supposed to perform atomic operations.

Signed-off-by: Ernesto Castellotti <erny.castell@gmail.com>

ErnyTech commented:
> I realized that the formatting has problems (badly converted tabs), I have to solve it

Fixed
